CONTRIBUTING.md instructions

mkfreeman opened the pull request:
This PR updates the instructions in the `CONTRIBUTING.md` file to more explicitly outline steps involved. The intention is to get people with less development experience empowered to make changes more easily. @zanarmstrong @ananya-roy I'd love to hear your thoughts on this -- if you have time to walk through the [updated instructions](https://github.com/observablehq/plot/blob/mkfreeman/contributing/CONTRIBUTING.md), that would be a very helpful assessment!

Fil commented on CONTRIBUTING.md:
```diff
@@ -2,21 +2,53 @@
-This should open http://localhost:8008/ where you can browse the tests.
+This should open http://localhost:8008/ in your browser where you can see the tests plots defined in `test/plots`. 
+
+![local deverlopment environment](img/localhost.png)
```
```suggestion
![local development environment](img/localhost.png)
```

Fil replied:
in img/localhost.png the names that are shown are the symbols exported by the js files in the test/plot/ folder. The file names are a bit different, and a few files (eg scales and legends) export multiple symbols.

Fil commented on CONTRIBUTING.md:
```diff
@@ -2,21 +2,53 @@
+
+![local deverlopment environment](img/localhost.png)
+
+If you'd like to create a new example as part of your development, you need to:
```
```suggestion
If you'd like to create a new test plot as part of your development, you need to:
```

these are not examples stricto sensu, but rather test plots that allow us to make sure that changes do not cause unwanted regressions in certain features. Examples are generally available in Observable notebooks, and (hopefully) convey best practices. Test plots are sometimes not good as examples.

Fil commented on CONTRIBUTING.md:
```diff
@@ -2,21 +2,53 @@
+2. Export the function from your `new-example.js` file in `test/index.js`. Because the first alphabetical option is displayed by on the local server, you may want to export the example as `_newExample`.
+
+### Development using Observable
+If you'd like to develop using Observable (or test your changes on an existing notebook), you'll neeed to first compile your changes into a single file:
```
```suggestion
If you'd like to develop using Observable (or test your changes on an existing notebook), you'll neeed to first bundle your changes into a single file:
```

Fil commented on CONTRIBUTING.md:
````diff
@@ -59,3 +91,10 @@ If your code intentionally changes some of the existing snapshots, simply blow a
+
+```bash
+# Push changes up to a new branch on GitHub
+git push -u origin user-name/some-feature
````
```suggestion
git push -u origin user-name/some-feature
```

I don't think this workflow is generally available. You should first fork the repo, then push to a new branch on your fork, and finally make a pull-request against the official repo.

mkfreeman replied:
This is interesting -- I wonder what a good way to callout different processes for our colleagues might be...

mbostock replied:
I’m not sure it’s appropriate to cover the basics of git, GitHub, terminal commands here. It might be better to link to the corresponding GitHub tutorials, e.g.,

https://docs.github.com/en/get-started/quickstart/fork-a-repo
https://docs.github.com/en/get-started/quickstart/contributing-to-projects

mbostock commented on CONTRIBUTING.md:
```diff
@@ -2,21 +2,53 @@
 git clone git@github.com:observablehq/plot.git
-cd plot
-yarn
+cd plot # cd into the folder
```
Does the comment “cd into the folder” help here, or does it just repeat what the code is saying? Is the expectation that the reader doesn’t know what “cd” does, and if so, should the comment explain that in English? It’s probably more appropriate to assume the reader knows what “cd” does, along with other terminal basics, here.

mkfreeman replied:
Good point, we can remove the annotation.

mbostock commented on CONTRIBUTING.md:
```diff
@@ -2,21 +2,53 @@
+If you'd like to create a new test plot as part of your development, you need to:
+
+1. Create a new `.js` file in `test/plots`, e.g., `test/plots/new-example.js` (an easy start is duplicating an existing example).
+2. Export the function from your `new-example.js` file in `test/index.js`. Because the first alphabetical option is displayed by on the local server, you may want to export the example as `_newExample`.
```
I would not recommend this, as then folks are likely to submit a pull request with this naming convention. We want the tests to be in alphabetical order in `test/plots/index.js`. Also, once you’ve selected your test, you get hot module reloading, so you don’t have to keep re-selecting your test as you develop. And, you can use the left and right arrow keys to flip through the tests quickly, as well as Enter + typing to quickly select your desired test via the keyboard.

mbostock replied:
The test file is `test/plots/index.js`, not `test/index.js`.

mkfreeman replied:
Noted about the naming convention (I hadn't used the hot-reloading enough to realize it kept your Select option). I'll remove that content.

mkfreeman replied:
And I'll update the screenshot :)

mbostock commented on CONTRIBUTING.md:
````diff
@@ -2,21 +2,53 @@
+
+```js
+// Load your copy into Observable
+Plot = require("http://localhost:8008/path/to/plot/dist/plot.umd.js")
````
I don’t think we should recommend this approach specifically (because it doesn’t work in Safari), and more generally, I think we should not recommend that people “develop in Observable”. The recommended development pattern is to use Vite locally as described above. I think it’s okay (desirable, even) to describe how to _share a demo_ on Observable of some in-development feature. And in that case, you’d be uploading your built plot.umd.js as a file attachment and saying something like:

```js
Plot = require(await FileAttachment("plot.umd.js").url())
```

mkfreeman replied:
I can appreciate that -- I'll remove this approach, and reframe the description about sharing on Observable.

ananya-roy reviewed (approved):
- Under Development, it says to export the function to `test/index,js`. I think you meant `test/plots/index.js`

- Are there guidelines for documenting changes (like a README or in some cases a published Observable notebook with a live example?) as part of submitting a PR?

mkfreeman commented:
Thanks @ananya-roy -- I've fixed the typo, and added a line about updating the `README.md` file (I'm not sure if there's any more detail we can provide as a general rule for updating the documentation).

mbostock reviewed (changes requested):
I want to reiterate some earlier feedback that was marked as resolved but was not addressed (at least, not explicitly).

The primary audience of CONTRIBUTING.md is the community (_i.e._ the public), not other Observable employees that can push directly to this repository. As such the recommended workflow for development, as @Fil pointed out, must be to fork this repository on GitHub rather than to clone this repository. In addition the {username}/{feature} branch naming convention need not apply; that is only needed because we, at Observable, by convention, share this repository.

mkfreeman commented:
Thanks for clarifying // resurfacing -- I've updated the Set Up instructions for the general public. I also switched to the default https code that GitHub has users copy (in case some doesn't have SSH keys set up). Happy to switch that back if desired.

mbostock commented on CONTRIBUTING.md:
```diff
@@ -2,21 +2,48 @@
 
 Observable Plot is released under the ISC license. You are welcome to fork this repository and to send us pull requests to contribute bug fixes or new features.
```
```suggestion
Observable Plot is released under the ISC license. You are welcome to [fork this repository](https://docs.github.com/en/get-started/quickstart/fork-a-repo) and to send us pull requests to contribute bug fixes or new features.
```

mbostock commented on CONTRIBUTING.md:
````diff
@@ -59,3 +86,8 @@ If your code intentionally changes some of the existing snapshots, simply blow a
 ```
+
+## Submitting a pull request
+Once you have completed your feature and confirmed that all test have passed using `yarn test`, you should push your changes up to GitHub and submit a [pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request).
````
We mention pull requests up above (I suggest we move this GitHub link above), so maybe this section could instead be more specific about our expectations for what makes a good pull request, or a good feature? Like, please include a motivating example, or why you think this feature will be useful. Or, please consider filing an issue, or checking for existing issues, before submitting a pull request, in case someone else has already thought about this topic. Or perhaps to start, this section could instead be called `## Documentation`, and delete this sentence?

mkfreeman replied:
I switched the last section to **Documentation** and moved the Observable instructions there.

mbostock commented on CONTRIBUTING.md:
```diff
@@ -59,3 +86,8 @@ If your code intentionally changes some of the existing snapshots, simply blow a
+## Submitting a pull request
+Once you have completed your feature and confirmed that all test have passed using `yarn test`, you should push your changes up to GitHub and submit a [pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request).
+
+Please make sure to update the `README.md` file to reflect changes to the library.
```
```suggestion
Please update the [README.md](./README.md) to reflect any changes to Plot’s public API.
```

mbostock commented on CONTRIBUTING.md:
```diff
@@ -1,22 +1,37 @@
+
+![local development environment](img/localhost.png)
+
+If you’d like to create a new test plot as part of your development, you need to:
+
+1. Create a new `.js` file in `test/plots`, e.g., `test/plots/new-example.js` (an easy start is duplicating an existing example).
+2. Export the function from your `new-example.js` file in `test/plots/index.js`. 
```
Isn’t this already covered below in the testing section, starting here?

> To add a new snapshot test, create a new JavaScript file in the `test/plots` folder.

mkfreeman replied:
Good point -- as someone less familiar with this structure of development, I wouldn't think to look in the `test` section for information about how to create a new example (so I've left a sentence directing them there).

mbostock commented:
I still want to make a bunch of edits, but for the sake of expediency, I’ll just make them directly rather than going more rounds on this review. Thank you for the contribution. 🙏

mbostock commented:
Apologies for the delay on this. I haven’t forgotten and will get to it soon.
